lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -44,73 +36,54 @@ def __init__(self, charm):
-            current_allocations = set(
-                self._charm.peers_data.get(self._scope, ALLOCS_TO_DELETE, "").split(",")
-            )
-            current_allocations.add(self._node.name)
-
-            self._charm.peers_data.put(
-                self._scope, ALLOCS_TO_DELETE, ",".join(current_allocations)
-            )
-
-    def cleanup(self) -> None:
-        """Delete all exclusions that failed to be deleted."""
-        need_voting_cleanup = self._charm.peers_data.get(self._scope, VOTING_TO_DELETE, False)
-        if need_voting_cleanup and self._delete_voting():
-            self._charm.peers_data.delete(self._scope, VOTING_TO_DELETE)
-
+    def allocation_cleanup(self) -> None:
+        """Delete alloc exclusions that failed to be deleted."""
         allocations_to_cleanup = self._charm.peers_data.get(
             self._scope, ALLOCS_TO_DELETE, ""
         ).split(",")
-        if allocations_to_cleanup and self._delete_allocations(allocations_to_cleanup):
+        if allocations_to_cleanup and self.delete_allocations_exclusion(allocations_to_cleanup):
             self._charm.peers_data.delete(self._scope, ALLOCS_TO_DELETE)
 
-    def _add_voting(self) -> bool:
-        """Include the current node in the CMs voting exclusions list of nodes."""
+    def add_voting(
+        self, alt_hosts: Optional[List[str]] = None, node_names: Optional[List[str]] = None
+    ) -> bool:
+        """Include the current node in the CMs voting exclusions list of nodes.
+
+        If node_names is not specified, then the current node is excluded for compatibility.
+        """
+        if set(node_names or []) == self._fetch_voting_exclusions(alt_hosts):
+            # no need to add the same nodes again
+            return True
+
+        add_nodes = "&node_names=" + (",".join(node_names) if node_names else self._node.name)
         try:
             self._opensearch.request(
                 "POST",
-                f"/_cluster/voting_config_exclusions?node_names={self._node.name}&timeout=1m",
-                alt_hosts=self._charm.alt_hosts,
+                "/_cluster/voting_config_exclusions?timeout=1m" + add_nodes,
+                alt_hosts=alt_hosts if alt_hosts else self._charm.alt_hosts,
                 resp_status_code=True,
                 retries=3,
             )
             return True
         except OpenSearchHttpError:
             return False
 
-    def _delete_voting(self) -> bool:
+    def delete_voting(self, alt_hosts: Optional[List[str]] = None) -> bool:
         """Remove all the voting exclusions - cannot target 1 exclusion at a time."""
         # "wait_for_removal" is VERY important, it removes all voting configs immediately
         # and allows any node to return to the voting config in the future
         try:
             self._opensearch.request(
                 "DELETE",
                 "/_cluster/voting_config_exclusions?wait_for_removal=false",
```

Comment:
I do not remember why did I set in the past `wait_for_removal=false` instead of `true`. I now don't see a reason why it shouldn't be `true`

Reply:
@Mehdi-Bendriss on my own tests, I could not get this DELETE to work with `wait_for_removal=true`... I wanted to move it to True as well, btw!

Reply:
@Mehdi-Bendriss, found this in the docs:

> Defaults to true, meaning that all excluded nodes must be removed from the cluster before this API takes any action.

So, wait_for_removal=true expects the node to be gone entirely... That is why it is not suitable for us.